### PR TITLE
OIDC logout feature for store

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Issue
Product-apim: https://github.com/wso2/product-apim/issues/4865

## Purpose
Users should be able to logged out of the application.

## Goals
User session in the browser should be destroyed.
User session in the IDP server should be destroyed.
User should be redirected to anonymous page once logged out.

## Approach
POST request to the oidc/logout endpoint will destory the IDP user logged in session. In this request id_token and redirection url will be provided. This id_token is generated when the user access token request to the /token endpoint happens and will be stored in the cookie. Once this POST request successfully completed, it will be redirected to provided auth/callback endpoint ,and in here, the user browser cookies will be removed, access token will be revoked and user will be redirected to the anonymous view.

## User stories
User should be able to successfully logged out of the application.

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/logout/logout.jag [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<%
+    var log = new Log("Logout Request Function");
+    var idTokenP1 = request.getCookie("AM_ID_TOKEN_DEFAULT_P1").value;
+    var idTokenP2 = request.getCookie("AM_ID_TOKEN_DEFAULT_P2").value;
+    var idToken = idTokenP1 + idTokenP2;
+    // var idToken = session.get('idToken');
+    var postLogoutRedirectURI = "https://localhost:9443/store-new/services/auth/callback";
+    var logoutState = "state_1";
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/logout/logout.jag]: @CrowleyRajapakse shall we use a more readable value here.

Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/logout/logout.jag]: yes, we can make this empty also, but it's provided to check the whether the logout is successful from IDP side, only if it's successful response will have query param state with value we have sent, we can check whether this logout is initiated by us not by another party, in the doc they have used value "state_1", we can use any value as well.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag [hunk 2]
```diff
@@ -1,64 +1,119 @@
...
+        var Base64 = org.apache.axiom.om.util.Base64;
+        var String = Packages.java.lang.String;
+        var Integer = Packages.java.lang.Integer;
+        var base64encoded = Base64.encode ( new String (clientId + ":" + clientSecret).getBytes());
+        var tokenEndpoint = "https://localhost:8243/token";
+        var result = post(tokenEndpoint, tokenRequestData,{"Authorization": "Basic " + base64encoded});
 
-    cookie = {'name': 'AM_REF_TOKEN_DEFAULT_P2', 'value': refresh_token_token_part_2, 'path': "/store-new/", "HttpOnly": true, "secure": true, "maxAge": -1};
-    response.addCookie(cookie);
+        log.debug("Token response: " + result.data);
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag]: @CrowleyRajapakse Shall we avoid logging token response data

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag [hunk 3]
```diff
@@ -1,64 +1,119 @@
...
+
+        var access_token_token_part_1 = access_token.substring(0, token_length/2);
+        var access_token_token_part_2 = access_token.substring(token_length/2, token_length);
+
+        var refresh_token = String(token_response.refresh_token);
+        token_length = token_response.refresh_token.length;
+        var refresh_token_token_part_1 = refresh_token.substring(0, token_length/2);
+        var refresh_token_token_part_2 = refresh_token.substring(token_length/2, token_length);
+
+        var response_data = {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag]: @CrowleyRajapakse  Shall we remove this `response_data` variable

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag [hunk 4]
```diff
@@ -1,64 +1,119 @@
 <%
     var log = new Log("Token Request Function");
+    var state = request.getParameter("state","UTF-8");
+    log.debug("state: " + state);
+    if ( state !== null ) {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store-new/services/login/token_callback.jag]: @ruwiniwj Can you please pull these improvements into your PR https://github.com/wso2/carbon-apimgt/pull/6466, Basically we would need to refactor the token_callback.jag file and split the login callback handler and logout callback handler into two separate files.
* Would need to add new routing in jaggery conf
* Would need to change the callback_url parameter in DCR call to accept two callback endpoints in login and logout requests

**Comment by malinthaprasan** [comment]
Closing this as this is included in the PR https://github.com/wso2/carbon-apimgt/pull/6466

**Pull request closed without merging**